### ISSUE TITLE
Static CDN Filters: Embedded Fonts result in Cross-Origin Request Blocked (CORS) error

### ISSUE THREAD

**Issue by jaswrks** [wpsharks/comet-cache#427]
Referencing: http://davidwalsh.name/cdn-fonts
For this reason, I suggest that we blacklist all font extensions by default. We can write a KB article to explain what is needed for fonts to be served cross-domain without issue, or just link to [this article](http://davidwalsh.name/cdn-fonts).

Referencing internal ticket: https://websharks.zendesk.com/agent/tickets/5283

**Comment by raamdev** [comment]
I'm labeling this as a bug since the current default configuration may result in this error message in the browser console:

> Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource

**Comment by jaswrks** [comment]
I was just reviewing this again. I was unable to reproduce an issue with `eot`, `woff`, `otf` or `ttf` files. The default set of whitelisted extensions in ZenCache Pro (powered by the WP Media Library) does not include these file types. So for that reason, unless a site owner explicitly enabled these files types, this should not be an issue in the current release.

Having said that. I suggest the following changes:
- Add `eot,ttf,otf,woff` to the default set of whitelisted extensions.
- Add `eot,ttf,otf,woff` to the default set of blacklisted extensions also.

This way enabling fonts requires only that you remove the extensions from your blacklist.

**Comment by jaswrks** [comment]
I've had this corrected. I'll submit a Pull Request once the repo changeover is done.

**Comment by raamdev** [comment]
Currently blocked by the repo changeover.

**Comment by raamdev** [comment]
This has been fixed in the latest development copy and will go out with the next release: 
- **Improvement** (Pro): Updated Static CDN Filters whitelist to exclude font file extensions `eot,ttf,otf,woff`, as they require custom CORS configurations to display properly. See [#427](https://github.com/websharks/zencache/issues/427).
- **Improvement** (Pro): Updated Static CDN Filters blacklist to include font file extensions `eot,ttf,otf,woff`, as they require custom CORS configurations to display properly. See [#427](https://github.com/websharks/zencache/issues/427).

**Comment by jaswrks** [comment]
> Add eot,ttf,otf,woff to the default set of blacklisted extensions also.

I think we goofed here. Fonts are most likely to be referenced by CSS, which is static. So when Static CDN Filters are applied, the CSS is getting moved to the CDN, and the fonts are then expected to live on the CDN too. By excluding them, we are actually creating a problem instead of solving one.

I think the reason I made this call before (i.e., goofed), is that when this change was made, the HTML Compressor was not picking up Static CDN Filters. We improved this later. So now they are, which means that CSS compressed by the HTML Compressor is now served from the CDN. Now that this has been updated, font file extensions can be included by default.

**Comment by raamdev** [comment]
@jaswsinc I just tested uploading a custom font and adding a custom `@font-face` declaration to my `style.css` file, then setting `.entry-title` to use that custom font face. 

I enabled ZenCache Pro v150709, HTML Compressor, and Static CDN Filters, then removed the list of fonts from the Static CDN Filters Blacklisted Extensions. Upon saving and reloading the front-end of the site (from a different, logged-out browser), I still received the Cross-Origin error and the font did not load (using Chrome 43 on Mac OS X).

![2015-07-15_16-16-17](https://cloud.githubusercontent.com/assets/53005/8708905/dc9875f6-2b0c-11e5-86e5-92c630c902ac.png)

![2015-07-15_16-14-51](https://cloud.githubusercontent.com/assets/53005/8708872/b3c5c7e6-2b0c-11e5-9170-23994a425192.png)

I proceeded to inspect the compressed CSS to confirm the URLs to the font files had been rewritten to point at the CDN and sure enough they had:

``` css
@font-face{font-family:'Raleway';src:url('http://d28ql8m1xkpbbr.cloudfront.net/wordpress/wp-content/themes/twentyfifteen/webfonts/raleway_thin-webfont.eot');src:url('http://d28ql8m1xkpbbr.cloudfront.net/wordpress/wp-content/themes/twentyfifteen/webfonts/raleway_thin-webfont.eot#iefix') format('embedded-opentype'),url('http://d28ql8m1xkpbbr.cloudfront.net/wordpress/wp-content/themes/twentyfifteen/webfonts/raleway_thin-webfont.ttf') format('truetype'),url('http://d28ql8m1xkpbbr.cloudfront.net/wordpress/wp-content/themes/twentyfifteen/webfonts/raleway_thin-webfont.woff') format('woff'),url('http://d28ql8m1xkpbbr.cloudfront.net/wordpress/wp-content/themes/twentyfifteen/webfonts/raleway_thin-webfont.svg#Raleway') format('svg');font-weight:normal;font-style:normal}
```

So, it would appear that it's not as simple as removing those font file extensions from the Blacklisted Extension on the Static CDN Filters...

---

---

This is what I added to the bottom of my `style.css` file to have it load the font I uploaded:

``` css
@font-face {
    font-family: 'Raleway';
    src:url('webfonts/raleway_thin-webfont.eot');
    src:url('webfonts/raleway_thin-webfont.eot?#iefix') format('embedded-opentype'),
        url('webfonts/raleway_thin-webfont.ttf') format('truetype'),
        url('webfonts/raleway_thin-webfont.woff') format('woff'),
        url('webfonts/raleway_thin-webfont.svg#Raleway') format('svg');
    font-weight: normal;
    font-style: normal;
}
```

Here's how I'm using the font:

``` css
.entry-title {
    font-family: "Raleway" !important;
}
```

Finally, [here's the Raleway font](https://www.theleagueofmoveabletype.com/raleway) in case you want to download and use exactly what I used to test.

**Comment by jaswrks** [comment]
Interesting. In your example the cross-domain restriction seems to be preventing the font files from being served from one domain to another, even when the CSS and the fonts are on the same domain. In your example, the site referencing the fonts is on another domain, and the restriction still kicks in.

I suppose that makes sense, given that browsers implement this restriction not as a security measure, but in support of font licensing; i.e., to help prevent fonts from being shared across hosts.

I haven't run through the same test that you did yet, but from what I'm seeing in your detailed report, it would seem that we need to do two things instead of one.
- Remove font files from the blacklist as you did.
- Add cross-domain headers to the WordPress `.htaccess` file whenever Static CDN Filters are enabled. See: https://www.maxcdn.com/one/tutorial/how-to-use-cdn-with-webfonts/

``` apache
<IfModule mod_headers.c>
  <FilesMatch "\.(ttf|ttc|otf|eot|woff|woff2|font.css|css|js)$">
    Header set Access-Control-Allow-Origin "*"
  </FilesMatch>
</IfModule>
```

_In theory, the CDN should pick up this header whenever it caches the fonts. Whenever the font is loaded from the CDN, the CDN will then be sending the cross-domain header just as the original site did. Noting that a CDN is supposed to pick up all headers that your site sends and mirror those automatically._

**Comment by raamdev** [comment]
@jaswsinc I added the following to my `.htaccess` file:

```
<IfModule mod_headers.c>
  <FilesMatch "\.(ttf|ttc|otf|eot|woff|woff2|font.css|css|js)$">
    Header set Access-Control-Allow-Origin "*"
  </FilesMatch>
</IfModule>
```

But I'm still getting the same error:

![2015-07-15_17-09-13](https://cloud.githubusercontent.com/assets/53005/8709963/3ee2d8da-2b14-11e5-82ec-cf286cd6da3a.png)

**Comment by jaswrks** [comment]
Did you try clearing the CDN cache also?

**Comment by jaswrks** [comment]
You might need to clear your own browser cache also.

**Comment by raamdev** [comment]
> Did you try clearing the CDN cache also?

Tried that. Same issue.

> You might need to clear your own browser cache also.

Tried that as well, the error persists.

**Comment by jaswrks** [comment]
Is there an `.htaccess` file in the directory where those fonts are stored? If so, does it include anything that would override the root `.htaccess` file and prevent the header from being sent? Have you confirmed that you do have `mod_headers` in your Apache install?

**Comment by raamdev** [comment]
> Have you confirmed that you do have mod_headers in your Apache install?

I'm testing this on `raam.websharks-inc.net`; doesn't that have `mod_headers`?

**Comment by raamdev** [comment]
> Is there an .htaccess file in the directory where those fonts are stored?

No, none. And no other `.htaccess` file anywhere that would be overriding it.

**Comment by jaswrks** [comment]
> I'm testing this on raam.websharks-inc.net; doesn't that have mod_headers?

Yes, it seems so: http://support.hostgator.com/articles/hosting-guide/hardware-software/apache-modules-that-are-on-all-of-our-shared-servers

You could remove the `<IfModule>` check to confirm. If you don't get a 500 internal server error with that check missing, `mod_headers` is in fact enabled.

``` apache
<FilesMatch "\.(ttf|ttc|otf|eot|woff|woff2|font.css|css|js)$">
    Header set Access-Control-Allow-Origin "*"
</FilesMatch>
```

---

UPDATE: I just noticed that Cloudfront doesn't mirror all headers by default. So you might be forced to configure this manually by adjusting your Cloudfront distribution; i.e., whitelist this header. We should probably look for articles on this topic as it pertains to each CDN and see if we can determine how difficult this will be for a novice to configure.

![2015-07-15_13-22-50](https://cloud.githubusercontent.com/assets/1563559/8710278/1cb7eed6-2af5-11e5-8f08-704e349eca40.png)

**Comment by raamdev** [comment]
> UPDATE: I just noticed that Cloudfront doesn't mirror all headers by default.

I tried editing my existing CloudFront distribution to add `Access-Control-Allow-Origin` to the whitelisted headers, but that did not seem to help.

So then I tried creating a new CloudFront distribution and configuring the whitelisted header during the distribution creation process. That seemed to work, as the Cross-Origin error went away. 

However, I then created a third CloudFront distribution, this time leaving the default (i.e., **Forward Headers = None**) and after updating my Static CDN Filters with the new CDN Hostname and re-testing, **I was unable to reproduce the Cross-Origin error**. 

So then I decided to remove the code from the `.htaccess` file. _I was still unable to reproduce the Cross-Origin error._ (I assume this was because of some sort of caching on the CDN side.)

Then, just to make sure I wasn't _really_ going crazy, I created a 4th CloudFront distribution, leaving the defaults (Forward Headers = None). After configuring the Static CDN Filters with the new CDN hostname and reloading the browser, I _was_ able to reproduce the Cross-Origin error (because now my `.htaccess` file was not setting the `Access-Control-Allow-Origin` header).

---

So the lesson here, I think, is that if your site is already experiencing the Cross-Origin error, it's not enough to simply start sending the `Access-Control-Allow-Origin` header in the `.htaccess` file. You must also recreate your CloudFront distribution and then reconfigure the Static CDN Filters with the new CDN Hostname.

On the plus side, **it appears that it's not necessary to configure the CDN to Forward Headers**. As long as I configure my `.htaccess` file to send the `Access-Control-Allow-Origin` header _before_ I configure the Static CDN Filters with a CDN hostname, things seem to work as expected (i.e., the `.htaccess` file does what it's supposed to).

It's likely that there's some sort of caching happening on the CDN side that is preventing an existing distribution from working properly _after_ updating the `.htaccess` file, but from a user-perspective (and from a documentation perspective), the above process is important to note. Also, when we start updating the `.htaccess` file, we'll want to make sure we do that _before_ there's any chance of any URLs being rewritten to use the CDN.

**Comment by jaswrks** [comment]
> On the plus side, it appears that it's not necessary to configure the CDN to Forward Headers.

Awesome! Wow, that sounds like it took some work.

> It's likely that there's some sort of caching happening on the CDN side that is preventing an existing distribution from working properly after updating the .htaccess file

Weird. When you tested this, how did you clear the CDN cache exactly? Did you invalidate the file on the Cloudfront side or alter the `iv` variable in the URL?

Hmm. Now that I'm looking this over again, I see what you posted earlier is missing the `?iv` variable entirely in these font URLs. I wonder if that's a bug in the HTML Compressor.

```
@font-face{font-family:'Raleway';src:url('http://d28ql8m1xkpbbr.cloudfront.net/wordpress/wp-content/themes/twentyfifteen/webfonts/raleway_thin-webfont.eot');src:url('http://d28ql8m1xkpbbr.cloudfront.net/wordpress/wp-content/themes/twentyfifteen/webfonts/raleway_thin-webfont.eot#iefix') format('embedded-opentype'),url('http://d28ql8m1xkpbbr.cloudfront.net/wordpress/wp-content/themes/twentyfifteen/webfonts/raleway_thin-webfont.ttf') format('truetype'),url('http://d28ql8m1xkpbbr.cloudfront.net/wordpress/wp-content/themes/twentyfifteen/webfonts/raleway_thin-webfont.woff') format('woff'),url('http://d28ql8m1xkpbbr.cloudfront.net/wordpress/wp-content/themes/twentyfifteen/webfonts/raleway_thin-webfont.svg#Raleway') format('svg');font-weight:normal;font-style:normal}
```

**Comment by raamdev** [comment]
@jaswsinc So what needs to happen here exactly to improve this in a future release? Here's what I have so far:
- [x] Remove `eot,ttf,otf,woff` from default Static CDN Filters Blacklisted Extensions (**Question:** how do we handle existing installs that already have this value? do we clear the value if only `eot,ttf,otf,woff`, i.e., it likely hasn't been modified?)
- [x] When enabling Static CDN Filters, ensure that the following is added to the root `.htaccess` file:
  
  ``` apache
  <IfModule mod_headers.c>
   <FilesMatch "\.(ttf|ttc|otf|eot|woff|woff2|font.css|css|js)$">
        Header set Access-Control-Allow-Origin "*"
   </FilesMatch>
  </IfModule>
  ```
- [x] If unable to write to `.htaccess` file to add the necessary code, show a Dashboard warning linking to a KB Article that explains what can happen without that code
- [x] Write and publish a KB Article that explains Cross-Origin Resource Sharing (CORS) and why its necessary to send an `Access-Control-Allow-Origin` header.

**Comment by raamdev** [comment]
> Weird. When you tested this, how did you clear the CDN cache exactly? Did you invalidate the file on the Cloudfront side or alter the iv variable in the URL?

Since there is no button or anything to invalidate the CDN cache in ZenCache (see #488), I've been simply changing the `iv` attribute name to something else (e.g., `iv1`, `iv2`, etc.).

**Comment by raamdev** [comment]
> I see what you posted earlier is missing the ?iv variable entirely in these font URLs. I wonder if that's a bug in the HTML Compressor.

Ah, yes, that could be it then! I pulled that line right out of the CSS file generated by the HTML Compressor.

**Comment by jaswrks** [comment]
> Ah, yes, that could be it then!

Confirmed. I opened an issue here: https://github.com/websharks/html-compressor/issues/62

**Comment by raamdev** [comment]
> Confirmed. I opened an issue here: websharks/html-compressor#62

Excellent! Thanks! I'll leave this issue `blocked` until that one is resolved.

**Comment by jaswrks** [comment]
> So what needs to happen here exactly to improve this in a future release?

Your list looks good to me. I would add that we need to wait for websharks/html-compressor#62 to be fixed also. I see you just did that by placing this on hold.

> (Question: how do we handle existing installs that already have this value? do we clear the value if only eot,ttf,otf,woff, i.e., it likely hasn't been modified?)

Yes, that would be my thought also. Seems like a good solution.

**Comment by raamdev** [comment]
> I'll leave this issue blocked until that one is resolved.

Actually, I don't see why I would need to. We can finish working on this even before the bug is resolved, correct?

**Comment by raamdev** [comment]
> I would add that we need to wait for websharks/html-compressor#62 to be fixed also. I see you just did that by placing this on hold.

Copy that. I'll leave this `blocked` for now.

**Comment by jaswrks** [comment]
Yes, I think we can start work. Sorry for the confusion. There are actually a few parts to this, so we could go ahead and tackle those that are ZenCache-specific.

## Suggested Next Actions
- Create a new sub-folder in [this directory](https://github.com/websharks/zencache-pro/tree/000000-dev/src/includes/templates) called: `htaccess`
- Create several files in that directory. One for each of the items that we want to add to the WordPress `.htaccess` file. We can start in this issue by adding one named: `templates/htaccess/cdn-filters.txt`. That file should contain:
  
  ``` apache
  <IfModule headers_module>
    <FilesMatch "\.(ttf|ttc|otf|eot|woff|woff2|font.css|css|js)$">
      Header set Access-Control-Allow-Origin "*"
    </FilesMatch>
  </IfModule>
  ```
- Add a new class member [here](https://github.com/websharks/zencache-pro/blob/150626/src/includes/closures/Plugin/InstallUtils.php#L194) called `$self->addWpHtaccess()` and that can be used to gather the templates that should be installed (based on plugin configuration) and pop those into the root `.htaccess` file for WordPress automatically.
- Add another class member: `$self->removeWpHtaccess()` to remove those entries.
- With those methods in place, make a call to `$self->addWpHtaccess()` here: https://github.com/websharks/zencache-pro/blob/150626/src/includes/closures/Plugin/InstallUtils.php#L17
- Make a call to `$self->addWpHtaccess()` here also: https://github.com/websharks/zencache-pro/blob/150626/src/includes/closures/Plugin/InstallUtils.php#L46
- Make a call to `$self->removeWpHtaccess()` here: https://github.com/websharks/zencache-pro/blob/150626/src/includes/closures/Plugin/InstallUtils.php#L65
- Make a call to `$self->removeWpHtaccess()` here also: https://github.com/websharks/zencache-pro/blob/150626/src/includes/closures/Plugin/InstallUtils.php#L87

---

Note that `addWpHtaccess()` should start by making a call to `removeWpHtaccess()` first. That way what it adds is from a fresh start; i.e., based on the current configuration of ZenCache.

**Comment by jaswrks** [comment]
I made a couple of edits to the above to correct some of the links and improve the instructions.

**Comment by raamdev** [comment]
> Suggested Next Actions

Thank you! I will assign this to myself and try to tackle that list tonight.

**Comment by jaswrks** [comment]
> (Question: how do we handle existing installs that already have this value? do we clear the value if only eot,ttf,otf,woff, i.e., it likely hasn't been modified?)

Yes, that would be my take also. I made a note of this a while back in the source code and left example code here that you could uncomment if you like. See: https://github.com/websharks/zencache-pro/blob/feature/347/src/includes/classes/VsUpgrades.php#L272-L276

**Comment by raamdev** [comment]
Noting that this issue is no longer `blocked` now that https://github.com/websharks/html-compressor/issues/62 has been closed.

**Comment by jaswrks** [comment]
## Nginx Equivalent

```
location ~* \.(?:ttf|ttc|otf|eot|woff|woff2|css|js)$ {
    add_header Access-Control-Allow-Origin *;
  }
```

**Comment by raamdev** [comment]
Drafted related KB Article: [Recommended NGINX Server Configuration](https://github.com/websharks/zencache-kb/issues/91)

**Comment by raamdev** [comment]
_Comment moved to https://github.com/websharks/zencache-pro/pull/143#issuecomment-148276628._

**Comment by raamdev** [comment]
@jaswsinc Can you confirm this all sounds correct?

[_Moved [here](https://github.com/websharks/zencache/issues/427#issuecomment-148869705)._]

**Comment by jaswrks** [comment]
That sounds absolutely correct to me.

> previous release in attempt to resolve

in "an" attempt

**Comment by raamdev** [comment]
Next Pro Release Changelog: 
- **Bug Fix**: Fixed a bug with Static CDN Filters and Cross-Origin Resource Sharing (CORS) that was generating a "Cross-Origin Request Blocked" error. ZenCache will now update the root `.htaccess` file to include a `Header set Access-Control-Allow-Origin "*"` rule for `ttf`, `ttc`, `otf`, `eot`, `woff`, `woff2`, `font.css`, `css`, and `js` files whenever the Static CDN Filters are enabled. This is necessary to avoid "Cross-Origin Request Blocked" errors. Note that if you are already experiencing this error, you should create and configure a new CDN hostname to resolve the issue. In our tests it appears that some CDNs are caching the initial header response received by the server, which means it's necessary to send the `Access-Control-Allow-Origin` header _before_ configuring the Static CDN Filters with a CDN hostname. See [Issue #427](https://github.com/websharks/zencache/issues/427#issuecomment-121774596).
- **Bug Fix**: Removed `eot,ttf,otf,woff` font extensions from the Static CDN Filter Blacklisted Extensions. These were added in a previous release in an attempt to resolve an issue with Cross-Origin Resource Sharing (CORS), however now that the HTML Compressor has been updated to work with Static CDN Filters, the CSS compressed by the HTML Compressor is now served from the CDN. Fonts are most likely to be referenced by CSS, which is static. So when Static CDN Filters are applied, the CSS is getting moved to the CDN and the fonts are then expected to live on the CDN too. By excluding them from the Static CDN Filter, we were creating a problem instead of solving one. This release removes the font extensions from the default Blacklisted Extensions so that fonts can be hosted on the CDN alongside the CSS that references them. See [Issue #427](https://github.com/websharks/zencache/issues/427#issuecomment-121777790).

**Comment by raamdev** [comment]
Referencing related KB Article: [What are Cross-Origin Request Blocked Errors?](http://zencache.com/kb-article/what-are-cross-origin-request-blocked-errors/)

**Comment by raamdev** [comment]
**ZenCache v151114** has been released and includes changes from this GitHub Issue. See the [v151114 announcement](http://zencache.com/zencache-pro-v151114-now-available/) for further details.

---

This issue will now be locked to further updates. If you have something to add related to this GitHub Issue, please open a new GitHub Issue and reference this one (#427).
